sdk/core/azure-core-amqp/src/main/java/com/azure/core/amqp/implementation/ReactorSession.java

```diff
@@ -349,33 +351,35 @@ public boolean removeLink(String linkName) {
-                provider.getReactorDispatcher().invoke(() -> {
-                    final LinkSubscription<AmqpReceiveLink> computed = openReceiveLinks.compute(linkName,
-                        (linkNameKey, existing) -> {
-                            if (existing != null) {
-                                logger.info("linkName[{}]: Another receive link exists. Disposing of new one.",
-                                    linkName);
-                                tokenManager.close();
+        return Mono.when(onActiveEndpoint(), tokenManager.authorize())
+            .then(Mono.create((Consumer<MonoSink<AmqpReceiveLink>>) sink -> {
+                try {
+                    // This has to be executed using reactor dispatcher because it's possible to run into race
+                    // conditions with proton-j.
+                    provider.getReactorDispatcher().invoke(() -> {
+                        final LinkSubscription<AmqpReceiveLink> computed = openReceiveLinks.compute(linkName,
+                            (linkNameKey, existing) -> {
+                                if (existing != null) {
+                                    logger.info("linkName[{}]: Another receive link exists. Disposing of new one.",
+                                        linkName);
+                                    tokenManager.close();
 
-                                return existing;
-                            }
+                                    return existing;
+                                }
 
-                            logger.info("connectionId[{}] sessionId[{}] linkName[{}] Creating a new receiver link.",
-                                sessionHandler.getConnectionId(), sessionName, linkName);
-                            return getSubscription(linkNameKey, entityPath, sourceFilters, receiverProperties,
-                                receiverDesiredCapabilities, senderSettleMode, receiverSettleMode, tokenManager);
-                        });
+                                logger.info("connectionId[{}] sessionId[{}] linkName[{}] Creating a new receiver link.",
+                                    sessionHandler.getConnectionId(), sessionName, linkName);
+                                return getSubscription(linkNameKey, entityPath, sourceFilters, receiverProperties,
+                                    receiverDesiredCapabilities, senderSettleMode, receiverSettleMode, tokenManager);
+                            });
 
-                    sink.success(computed.getLink());
-                });
-            } catch (IOException | RejectedExecutionException e) {
-                sink.error(e);
-            }
-        }));
+                        sink.success(computed.getLink());
+                    });
+                } catch (IOException | RejectedExecutionException e) {
+                    sink.error(e);
+                }
+            }))
+            .doOnError(t -> tokenManager.close());
```

Comment:
I was trying to understand when to use doOnError and onErrorResume, but in this case, we probably want to continue publishing the error downstream, just this is another side-effect, to clean up?

Reply:
I often have trouble discerning when to use one operator versus the other. But this was helpful.
https://stackoverflow.com/questions/58167842/the-difference-between-onerrorresume-and-doonerror

Reply:
updated to use `onErrorResume` thanks for the feedback!